### PR TITLE
[SPARK-35912][SQL] Fix cast struct contains null value to string/struct

### PR DESCRIPTION
**Pull request by cfmcgrady**
### What changes were proposed in this pull request?

This PR fixes an issue that cast the struct which contains null value to other type has a difference result when we enable/disable codegen.

Here is an example:

```scala
  val schema = StructType(Seq(StructField("value",
    StructType(Seq(
      StructField("x", IntegerType, nullable = false),
      StructField("y", IntegerType, nullable = false)
    ))
  )))
  schema.printTreeString()
  // root
  // |-- value: struct (nullable = true)
  // |    |-- x: integer (nullable = false)
  // |    |-- y: integer (nullable = false)
  
  val testDS = Seq("""{"value":{"x":1}}""").toDS
  val jsonDS = spark.read.schema(schema).json(testDS)

  jsonDS.show()
  // incorrect result
  // +---------+
  // |    value|
  // +---------+
  // |{1, null}|
  // +---------+

  spark.sql("set spark.sql.codegen.wholeStage=false")
  jsonDS.show()
  // the result we expected due to y.nullable = false.
  // +------+
  // | value|
  // +------+
  // |{1, 0}|
  // +------+
```
Actually, the result should be depending on the field nullable setting. this bug also happens when we cast struct to struct.

### Does this PR introduce _any_ user-facing change?

No, only bug fix.

### How was this patch tested?

New test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -1098,29 +1100,37 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
...
       buffer: ExprValue,
       ctx: CodegenContext): Block = {
-    val structToStringCode = st.zipWithIndex.map { case (ft, i) =>
-      val fieldToStringCode = castToStringCode(ft, ctx)
-      val field = ctx.freshVariable("field", ft)
-      val fieldStr = ctx.freshVariable("fieldStr", StringType)
-      val javaType = JavaCode.javaType(ft)
-      code"""
-         |${if (i != 0) code"""$buffer.append(",");""" else EmptyBlock}
-         |if ($row.isNullAt($i)) {
```
Comment (cfmcgrady) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1112]: When the actual value is null, for primitive type field, `row.isNullAt(i)` return ture, but `row.getXXX` return a default value.

For exmaple:

```scala
val r = new org.apache.spark.sql.catalyst.expressions.GenericInternalRow(Array(1, null))
println(r.getInt(0))   // 1
println(r.getInt(1))   // 0
println(r.isNullAt(1)) // true
```

so we cann't only check `row.isNullAt(i)` here, we need to do the same logical like  `BoundReference.doGenCode()`, add nullable check.

**Comment by HyukjinKwon** [comment]
how is the cache issue related to the cast?

**Comment by cfmcgrady** [comment]
> how is the cache issue related to the cast?

HI, @HyukjinKwon 
The root cause is cast struct to other types, updated the pr description.

**Comment by cfmcgrady** [comment]
cc @cloud-fan @viirya @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -406,19 +406,21 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
         if (row.numFields > 0) {
           val st = fields.map(_.dataType)
           val toUTF8StringFuncs = st.map(castToString)
-          if (row.isNullAt(0)) {
+          if (fields(0).nullable && row.isNullAt(0)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:409]: if `fields(0).nullable` is false, how can `row.isNullAt(0)` be true?

**Comment by HyukjinKwon** [comment]
Hey mind explaining why cast path issue is related to being cached?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:409]: (I have the same question)

Comment (cfmcgrady) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:409]: If user create dataframe from `spark.internalCreateDataFrame()`, the `row.isNullAt()` may be null even though the schema nullable is false.
For instance:
```scala
  val schema = StructType(Seq(
    StructField("x",
      StructType(Seq(
        StructField("y", IntegerType, true),
        StructField("z", IntegerType, false)
      )))))
  val rdd = spark.sparkContext.parallelize(Seq(InternalRow(InternalRow(1, null))))
  val df = spark.internalCreateDataFrame(rdd, schema)
  df.show
  // current master branch output
  //  +---------+
  //  |        x|
  //  +---------+
  //  |{1, null}|
  //  +---------+
```

Although the `spark.internalCreateDataFrame()` is sql package private API, but `spark.read.json()` and `spark.read.csv()` call it without null value handled.(the example show in pr description)

**Comment by cfmcgrady** [comment]
> Hey mind explaining why cast path issue is related to being cached?

Actually, the cached result is what we want. The issue is that `spark.read.json` is not an expected result when the input schema has a nullable setting to false. From the log, we found it's a cast operation.
```
=== Result of Batch Preparations ===
 CollectLimit 21                                    CollectLimit 21
!+- Project [cast(value#5 as string) AS value#18]   +- *(1) Project [cast(value#5 as string) AS value#18]
!   +- Scan ExistingRDD[value#5]                       +- *(1) Scan ExistingRDD[value#5]
```

**Comment by HyukjinKwon** [comment]
Shouldn't it fail instead of setting it as `0`? I feel like the handling should be done somewhere in `JacksonParser`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:409]: Then we need to fix the nullability. There are so many places in the Spark codebase that relies on nullability to do optimizations. It's not possible to change all of them to not trust the nullability anymore.

Can we fix `spark.read.json()` to set the nullability correctly?

Comment (cfmcgrady) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:409]: Ok, let me try.

**Comment by cfmcgrady** [comment]
> Shouldn't it fail instead of setting it as `0` or `null`? I feel like the handling should be done somewhere in `JacksonParser`.

Thanks for your suggestion, I'll try.

**Comment by cfmcgrady** [comment]
Create a new [PR](https://github.com/apache/spark/pull/33212)

BTW, Shall we merge this PR? The cast issue may occur when the user create dataframe from API `spark.internalCreateDataFrame()`. cc @cloud-fan @maropu

**Comment by cloud-fan** [comment]
`spark.internalCreateDataFrame()` is a private API and users should be responsible if they do something wrong. We can't merge this PR, because we can't sacrifice the performance in the happy path for invalid usage of private APIs from users.

**Pull request closed without merging**